src/net_processing.cpp

```diff
@@ -3969,43 +3970,39 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
     return;
 }
 
-bool PeerManagerImpl::MaybeDiscourageAndDisconnect(CNode& pnode)
+bool PeerManagerImpl::MaybeDiscourageAndDisconnect(CNode& pnode, Peer& peer)
 {
-    const NodeId peer_id{pnode.GetId()};
```

Comment:
Indeed it would, but minimizing lines changed isn't the right metric to judge a commit. Instead we should aim to maximize the clarity and expressiveness of the code after the change. If I was writing this from scratch, I wouldn't create a local copy of the peer id just to use in log lines.

Reply:
Hmm, actually I think I disagree with that (well, at least after exaggerating it into a bigger principle) -- review is a much more important bottleneck than just about anything else, so optimising the code to be easier to review is much more helpful than optimising it to make perfect sense if we were going to erase our git history and anonymously post the current tree to a mailing list. I think just about any time there's a choice between correct code that's easier to review and correct code that's better according to some other aesthetic, easier to review should be the winner. At least until review is no longer a serious bottleneck for the project.

Obviously, `peer_id` to `peer.m_id` is trivial, but if we're talking what we should aim to maximise, I think ease of review is the single right answer. I don't think you'll often find that being far off from optimising for clarity, though. (I don't think "expressiveness" applies either way tbh)

Reply:
This discussion is a bit abstract, but I'm going to stand by my assertion. We should always be aiming to improve the quality of the code by standardizing to a clear and consistent style and by taking advantage of higher-level abstractions. One example: if I find a hand written loop that I need to modify, I may well replace it with an stl algorithm. That's undoubtedly a bit more effort to review, but results in higher-quality code. If we always aim to minimize deltas in code changes, then we'll only ever converge to local maxima of quality.

Reply:
The point of review being a bottleneck is that we won't ever progress to a global maximum of quality -- if we could do that, by definition we wouldn't be bottlenecked. Every merge is a reflection of that review bottleneck: did we spend our limited review cycles on working out if a while loop does the same thing as the for loop it replaces, or did we spend it on working out if muhash or minisketch does what it claims on the box? If every PR spends 5% of review time checking that shuffling code around is correct, then that's 1-in-20 fewer PRs we can make that affect people that don't see the code. Or it's an incentive to get in the habit of merging code that's not thoroughly reviewed in order to avoid delaying things because we've run out of review resources, which is worse.

That doesn't mean we shouldn't always be aiming to improve the quality of the code; just that we should be spending the resources we have on the changes that are most valuable, even if that means leaving some good things for (possibly much) later.

(Personally, I find most of the STL algorithms worse than the loops they replace -- range based for loops in c++ are pretty pleasant and behave pretty sensibly; for me, dealing explicitly with iterators and lambdas generally isn't pleasant -- they're harder to understand and easier to make mistakes with, for at best a minimal performance benefit, and for me at least any safety benefits are offset by the lack of clarity. There are more important battles to fight, and I'm hoping that perhaps familiarity might eventually breed affection, but for now they certainly don't seem like moving toward even a local maximum to me)

Reply:
Thanks AJ, these are good points.

You're wrong about stl algorithms though :grin: